### PR TITLE
Update pod status once node becomes NotReady

### PR DESCRIPTION
**Pull request by freehan**
Update pod ready status to false once its node gets into NotReady status. This is to avoid bad service endpoint due to node down.

### CODE REVIEW

**Comment by freehan** [comment]
Related to #17284 
@thockin

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -746,6 +748,38 @@ func (nc *NodeController) deletePods(nodeName string) (bool, error) {
 	return remaining, nil
 }
 
+// update ready status of all pods running on given node from master
+// return true if success
+func (nc *NodeController) updatePods(nodeName string, readyStatus api.ConditionStatus) (bool, error) {
+	glog.Infof("Update Ready Status of pods on node [%v]", nodeName)
```
Comment (thockin) [inline pkg/controller/node/nodecontroller.go]: this should probably be V(2)

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -746,6 +748,38 @@ func (nc *NodeController) deletePods(nodeName string) (bool, error) {
 	return remaining, nil
 }
 
+// update ready status of all pods running on given node from master
+// return true if success
+func (nc *NodeController) updatePods(nodeName string, readyStatus api.ConditionStatus) (bool, error) {
```
Comment (thockin) [inline pkg/controller/node/nodecontroller.go]: given that this is called on true-> false but not false->true, maybe this should be less generically named and parameterized?  setPodsNotReady(node) or something?

File: pkg/controller/node/nodecontroller_test.go [hunk 3]
```diff
@@ -597,6 +597,157 @@ func TestMonitorNodeStatusUpdateStatus(t *testing.T) {
 	}
 }
 
+func TestMonitorNodeStatusUpdatePodStatus(t *testing.T) {
```
Comment (thockin) [inline pkg/controller/node/nodecontroller_test.go]: What is "Monitor" - not present in the code being tested.

**Comment by thockin** [comment]
Just naming nits from me.  I'll be OOO rest of this week.

**Comment by thockin** [comment]
oh yeah, this should have a real end-to-end test.  @ArtfulCoder is the king of concise e2es.

File: pkg/controller/node/nodecontroller.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: +1 agree, maybe ```markAllPodsNotReady```

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -428,6 +429,7 @@ func (nc *NodeController) monitorNodeStatus() error {
 			// Report node event.
 			if readyCondition.Status != api.ConditionTrue && lastReadyCondition.Status == api.ConditionTrue {
 				nc.recordNodeStatusChange(node, "NodeNotReady")
+				nc.updatePods(node.Name, api.ConditionFalse)
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: the response from ```updatePods``` appears to be ignored.  why?

File: pkg/controller/node/nodecontroller.go [hunk 5]
```diff
@@ -746,6 +748,38 @@ func (nc *NodeController) deletePods(nodeName string) (bool, error) {
...
+			continue
+		}
+
+		for i, cond := range pod.Status.Conditions {
+			if cond.Type == api.PodReady {
+				pod.Status.Conditions[i].Status = readyStatus
+				pod, err := nc.kubeClient.Pods(pod.Namespace).UpdateStatus(&pod)
+				if err != nil {
+					glog.Warningf("Failed to updated status for pod %q: %v", kubeletutil.FormatPodName(pod), err)
+					return false, err
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: can this return an aggregated error instead?

rather than breaking out here and returning false directly, can this just ```continue``` so each pod is visited?

right now it looks like if one pod update fails, the others are not updated which is not ideal.

Comment (freehan) [inline pkg/controller/node/nodecontroller.go]: Good point. I will make the fix.

File: pkg/controller/node/nodecontroller.go (hunk 1 above)
Comment (freehan) [inline pkg/controller/node/nodecontroller.go]: Fixed

File: pkg/controller/node/nodecontroller.go (hunk 2 above)
Comment (freehan) [inline pkg/controller/node/nodecontroller.go]: Fixed

File: pkg/controller/node/nodecontroller_test.go (hunk 3 above)
Comment (freehan) [inline pkg/controller/node/nodecontroller_test.go]: Because the unit test is triggering monitorNodeStatus function, and this function triggers the pod mark down.

File: pkg/controller/node/nodecontroller.go (hunk 4 above)
Comment (freehan) [inline pkg/controller/node/nodecontroller.go]: Will add it.

**Comment by davidopp** [comment]
I think the fact that you only transition the pods in one direction (from pod.Status.Conditions["Ready"].Status == api.ConditionTrue to api.ConditionFalse) is problematic because the node itself may transition back from NodeReady==api.ConditionUnknown or NodeReady==api.ConditionFalse to NodeReady==api.ConditionTrue without evicting any pods if the node becomes reachable (from ConditionUnknown) or ready (from ConditionFalse) within five minutes of the forward transition.

This means you will "strand" pods in non-ready state even though the machine has gone back to NodeReady==api.ConditionTrue and has no plan of evicting them.

The fix is to transition the pods back to pod.Status.Conditions["Ready"].Status = api.ConditionTrue if the node transition to NodeReady==api.ConditionTrue.

It should be easy to add the code to do this in monitorNodeStatus() in the same place as you made the change in the current version of the PR.

**Comment by freehan** [comment]
@davidopp 
If the node came back, kubelet will sync the pod status. So the nodecontroller does not need to act. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L322

Any other corner case you can think of?

**Comment by freehan** [comment]
Found a corner case. When network partition happens between a node and nodecontroller. Assuming kubelet is still running on node. Based on this PR, After 40s of nodeMonitorGracePeriod, nodecontroller will mark the node and its pods as NotReady. If the connectivity recovers shortly after, the pods will stay NotReady. The reason is when kubelet syncs pod status, first it calculates the pod status and compare it against the local status cache. If they are the same, then do nothing. In this case, the pod status in cache is ready but it is notReady on master side. 

To Do:
1. Add a nodedown e2e test
2. Sync the pods status when network recovers on kubelet #8818

**Comment by derekwaynecarr** [comment]
please poke when this is ready for another review.

**Comment by freehan** [comment]
@derekwaynecarr @thockin @davidopp 
Dependency has been merged. This PR is ready for review.

**Comment by bprashanth** [comment]
Ping original reviewers. I'll do a first pass shortly, and plan to merge to help clear the water with e2e failures.

**Comment by bprashanth** [comment]
@derekwaynecarr (I think the other two are ooo)

File: pkg/controller/node/nodecontroller.go [hunk 6]
```diff
@@ -832,6 +836,39 @@ func (nc *NodeController) deletePods(nodeName string) (bool, error) {
...
+		}
+
+		for i, cond := range pod.Status.Conditions {
+			if cond.Type == api.PodReady {
+				pod.Status.Conditions[i].Status = api.ConditionFalse
+				glog.V(2).Infof("Updating ready status of pod %v to false", pod.Name)
+				pod, err := nc.kubeClient.Pods(pod.Namespace).UpdateStatus(&pod)
+				if err != nil {
+					glog.Warningf("Failed to updated status for pod %q: %v", format.Pod(pod), err)
+					complete = false
```
Comment (bprashanth) [inline pkg/controller/node/nodecontroller.go]: get rid of the bool return value?
```go
errMsg := []string{}
...
if err != nil {
  errMsg = append(fmt.Sprintf("%v", err))
}
...
if len(errMsg) == 0 {
  return nil
}
return fmt.Errorf("%v", strings.Join(errMsg, ", "))
```

File: test/e2e/resize_nodes.go [hunk 7]
```diff
@@ -388,6 +396,17 @@ func performTemporaryNetworkFailure(c *client.Client, ns, rcName string, replica
 	// network traffic is unblocked in a deferred function
 }
 
+// shut down eth0 on node for 180 seconds to simulate temporary network partition
+func inititateTemporaryNetworkPartition(c *client.Client, node *api.Node) {
+	host := getExternalIPforSSH(node)
+	Logf("Turning off network interface on %s", node.Name)
+	partitionCmd := "nohup sh -c 'sleep 10 && sudo ifdown eth0 && sleep 180 && sudo ifup eth0' >/dev/null 2>&1 &"
+	if result, err := SSH(partitionCmd, host, testContext.Provider); result.Code != 0 || err != nil {
+		LogSSHResult(result)
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Can you just use `Expect(issuSSHCommand("nohup sh", testContext.Provider, node)).NotTo(HaveOccurred())` ?

File: test/e2e/resize_nodes.go [hunk 8]
```diff
@@ -560,6 +579,53 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+				opts := api.ListOptions{FieldSelector: fields.OneTermEqualSelector(client.PodHost, node.Name)}
+				if err = waitForMatchPodsCondition(c, opts, "Running and Ready", podReadyTimeout, podRunningReady); err != nil {
+					Failf("Pods on node %s are not ready and running within %v: %v", node.Name, podReadyTimeout, err)
+				}
+
+				By(fmt.Sprintf("Performing temporary network partition on node %s", node.Name))
+				inititateTemporaryNetworkPartition(c, &node)
+
+				By("Observing node and pod status change from Ready to NotReady after network partition")
+				if !waitForNodeToBe(c, node.Name, api.NodeReady, false, resizeNodeNotReadyTimeout) {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: This is racy, the machine this test is running on could be overloaded and the ifup script will come back up in 180s. Can you handle this another way? What you should do is:
* Setup a watch in your test
* Network partition node
* Wait for node to come up
* Go through the list of watch notifications returned and confirm you got a not ready notification

File: test/e2e/resize_nodes.go [hunk 9]
```diff
@@ -560,6 +579,53 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+					Failf("Node %s did not become NotReady within %v", node.Name, resizeNodeNotReadyTimeout)
+				}
+				if err = waitForMatchPodsCondition(c, opts, "NotReady", podNotReadyTimeout, podNotReady); err != nil {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: If this is the only place you're using podNotReady, I'd rather have an inline  
```
func(p *api.Pod) (bool, error){ if podReady()...} 
```

File: test/e2e/resize_nodes.go [hunk 10]
```diff
@@ -560,6 +579,53 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+					Failf("Pods on node %s did not become NotReady within %v: %v", node.Name, podNotReadyTimeout, err)
+				}
+
+				//Wait for network connectivity to recover and check node/pod status
+				time.Sleep(30 * time.Second)
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Is this sleep necessary?

File: test/e2e/util.go [hunk 11]
```diff
@@ -541,6 +552,33 @@ func waitForPodCondition(c *client.Client, ns, podName, desc string, timeout tim
 	return fmt.Errorf("gave up waiting for pod '%s' to be '%s' after %v", podName, desc, timeout)
 }
 
+// waitForMatchPodsCondition finds match pods based on the input ListOptions.
+// waits and checks if all match pods are in the given podCondition
+func waitForMatchPodsCondition(c *client.Client, opts api.ListOptions, desc string, timeout time.Duration, condition podCondition) error {
```
Comment (bprashanth) [inline test/e2e/util.go]: Can you just reuse the pattern/functions here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L2258?

**Comment by bprashanth** [comment]
@freehan I want to get this in soon to see if it will help with https://github.com/kubernetes/kubernetes/issues/19665, so i suggest sending a follow up pr with the e2e and keeping this pr to marking nodes+unittests.

**Comment by freehan** [comment]
@bprashanth  Simplified this PR to one commit.

File: pkg/controller/node/nodecontroller.go [hunk 12]
```diff
@@ -489,6 +490,9 @@ func (nc *NodeController) monitorNodeStatus() error {
 			// Report node event.
 			if readyCondition.Status != api.ConditionTrue && lastReadyCondition.Status == api.ConditionTrue {
 				nc.recordNodeStatusChange(node, "NodeNotReady")
+				if err = nc.markAllPodsNotReady(node.Name); err != nil {
```
Comment (bprashanth) [inline pkg/controller/node/nodecontroller.go]: Not sure why this method is in a poll loop instead of triggered by a watch (not your problem, just saying), there's a chance it will still not realize as soon as a node falls off the netowork so the reboot test (the one that needs the pods to become not ready) can still fail.  How often does this poll loop run?

**Comment by bprashanth** [comment]
LGTM thanks, will wait to see if cq goes green. if it fails because of the service test again (https://github.com/kubernetes/kubernetes/issues/19665) I might hand merge this.

File: pkg/controller/node/nodecontroller.go (hunk 12 above)
Comment (freehan) [inline pkg/controller/node/nodecontroller.go]: Looks like it runs every 5 seconds... The timeout for node notready is 40 secs. So the worst case is 45 secs to flip node to notready.

**Comment by freehan** [comment]
ref: #13346

**Comment by zaa** [comment]
Guys, could you please include the fix into 1.1x branch? I do not see it in the code tagged with v1.1.8-beta.0

@freehan

**Comment by bprashanth** [comment]
@freehan @roberthbailey cherry-pick? 

tl;dr to judge worthiness: this essentially marks the pod as NotReady if the node becomes NotReady. Previously we'd have to wait 5m for the node controller timeout, when it would just delete the pods. With this, in the 40s (kubelet ready timeout) -> 5m period, the pod will at least leave the service, and when 5m hits if the node is still NotReady the nodecontroller will delete the pod anyway.

**Comment by thockin** [comment]
This depended on other changes that I think add up to an overly-large
change.  I'll let @freehan decide, but that's my inclination.

**Comment by freehan** [comment]
@zaa This PR depends on some changes in kubelet, which is undergoing refactoring.

 Plus, 1.2 will be out soon. Can you wait for 1.2?

**Comment by zaa** [comment]
@freehan ok, sure. Thank you for the explanation.
